### DISCUSSION TITLE
Comparison of horde-ad with writing gradients manually using hmatrix

### DISCUSSION THREAD

**Discussion by Mikolaj** [Mikolaj/mostly-harmless#19] [General]
Let's gather in this discussion all materials that compare horde-ad with handwritten gradients and let's discuss here. Let's start by attaching the results of experiments with fully connected neutral networks for MNIST. Simpler experiments may be required to answer some of the harder questions.

The comparison can include at least: 

1. comparison of our 'Delta`-trace expressions with hand-written gradient expressions
2. comparison of the exact mathematical functions they encode, when interpreted as gradients
3. comparison of performance: loss/iteration
4. comparison of performance: loss/wall-time or, equivalently given point 3., iteration/wall-time
5. numerical stability

Here is a copy of SPJ's notes from 2 March meeting posing these and related questions:

* Compare the delta expression from the trace of a fully connected neural network for MNIST. Mikolaj showed side by side.
* Broadly order is the same, but very different things in the middle.  e.g. one matrix-vector multiply (manual version) turns into two in the auto version; and an outer-product in manual turns into nothing in auto.
* Training quality seems similar in both.
* Seems to compute same thing in a different way -- we are faster than the manual one!
* Question: do we generate the same derivatives?   Or different?
* Question: are we actually computing the same function (i.e. we can algebraically transform one into the other), but more efficiently?  Could we test whether we are computing the same function?
* Question: what is the smallest example that shows different code between manual and automatic?   What thought process went into the manual version?  Can we understand why our system generates a different algebraic expression than the manual version?
* Stability: MK says that "with Floats the network doesn't train; with Doubles it does".   (For some experiments; not MNIST.)
The relevant paper about numerical stability of AD is this one: https://link.springer.com/content/pdf/10.1007/s00607-011-0162-z.pdf
* Question: can we make a like-for-like comparison, rather than "manual version on right, simple; but on left is Delta version and you have to execute it backwards".

**Comment by Mikolaj** [comment]
Here's a screenshot of, side by side, delta expressions from the trace of a fully connected neural network for MNIST (for brevity, instead of 500 and 150 neuron hidden layers, 4 and 3 neuron hidden layers, but delta constructors don't change) vs the analogous handwritten gradient using hmatrix:

![image](https://user-images.githubusercontent.com/281893/154974058-6206dd47-9a25-4030-8cf2-927a97e2362b.png)

The original unedited printout of the delta expression is at https://gist.github.com/Mikolaj/a0e1086b711bff5dea3c97667bb93f6c.

This shows the delta expression bindings in the order corresponding to that of the handwritten gradient expressions. And old version of the screenshot is at https://github.com/Mikolaj/mostly-harmless/discussions/16?sort=new#discussioncomment-2200709. It shows the delta expressions in the reverse order (and old syntax), which corresponds to the order in the example in the paper. It's based on `let` expressions, while the screenshot included above uses `where` expressions (definitions given below their use).

**Comment by Mikolaj** [comment]
Graphs with X=wall time, Y=training error for the fully connected neural network for MNIST. This is 2 epochs, each contains randomly shuffled 60000 batches of sgd, each batch has only 1 sample (that's the original backprop setting that I'm comparing against). The artefact at the end and the X axis starting at negative 500 seconds is just me ignorantly flailing with (an old version) of gnuplot. I slightly doctored the data, too, to ensure the y axis extent is the same in both plots (not the x axis, sadly). First, the plot of training using the manual gradient: 

![image](https://user-images.githubusercontent.com/281893/155859317-9dffe94f-d5fb-40aa-b71a-82f8dc013f0a.png)

and then the plot of training with horde-ad automatically derived gradient for the same hmatrix-based forward computation:

![image](https://user-images.githubusercontent.com/281893/155859905-5bc798e4-d650-477b-8945-a6e2342dd8f4.png)

As a bonus and for comparison with something that doesn't train (in reasonable time at least), the same plot for exactly the same scenario, but using the "forward gradient" from "Gradients without Backpropagation" by Atilim Gunes Baydin, Barak A. Pearlmutter, Don Syme, Frank Wood and Philip Torr. 

![image](https://user-images.githubusercontent.com/281893/156019226-231543be-7e93-48ef-b5ae-e66511d52da0.png)

As seen, it doesn't train or trains very slowly. Please disregard the absolute time values, because the implementation I'm using is naive and slow, using the straightforward semantics of our delta expressions). However, the number of iterations is exactly the same as in the graphs above. 

After adding mini-batches, learning rate decay and a much lower initial learning rate, all as in the paper, it does train, though still very slowly (e.g., the initial fragments looks very much like the whole graph above, only a little more smooth thanks to mini-batches). The very MNIST neutral network they are using in the paper, which is not 500x150 hidden neurons, but 1024x1024 and uses relu instead of logistic for activation in the hidden levels, produces NaN very quickly, so they must be using some normalization/regularization/gradient clipping, which they are not describing in the paper (or my implementation has a bug, but exploding gradient with relu is nothing unusual and it works fine without relu). Therefore I can't reproduce their experimental results.

Edit: I think the comparative slowness to train of the version of the nn that converges at all is due to less aggressive following the gradient, which is necessary to prevent the divergence caused by the extra noise from randomness inherent in the method. I suppose this can often be fine-tuned so that the slowdown is not so extreme and still the nn converges. However, I didn't do any fine tuning, but just copied the settings from the paper (but applied to my nn that is much smaller and uses logistic activation instead of relu, so this may be additionally non-optimal). OTOH, analogous graphs from the paper do show very slow convergence and break off before the loss is lowered to acceptable levels.
